Change Creative Buttons Texture

matthewperiut opened the pull request:
# Screenshots of updated texture and position

<img width="400" alt="image" src="https://github.com/user-attachments/assets/7098faaf-90ce-4fdc-9292-2a9a1e7b9e43" />
<img width="401" alt="image" src="https://github.com/user-attachments/assets/0aac336e-499a-424c-9cd3-09b5a41c0240" />
<img width="403" alt="image" src="https://github.com/user-attachments/assets/3c6a41f1-b178-4af9-887a-d7c39b9491ce" />
<img width="407" alt="image" src="https://github.com/user-attachments/assets/3cdc195c-8e1c-4296-a3b9-69d00263d08f" />

Programmer art resource pack:
<img width="404" alt="image" src="https://github.com/user-attachments/assets/58b82c32-b76d-4a2c-9be4-a39329a3198e" />  
<img width="401" alt="image" src="https://github.com/user-attachments/assets/b71d2d95-895e-48b0-b256-a9ae2643afb3" />  

# **BELOW IS OUTDATED!**

![image](https://github.com/user-attachments/assets/05abf95f-29b4-40ab-befe-9a3546e3b62a)

![2024-11-11_17 20 15](https://github.com/user-attachments/assets/aed32684-fe1f-45a5-9a11-2a8b382272b5)
![2024-11-11_17 20 18](https://github.com/user-attachments/assets/3a4304d8-1873-47c2-ab5a-df10eb570725)
![2024-11-11_17 20 21](https://github.com/user-attachments/assets/78fa60f4-1c69-4879-b551-95747d10eea6)
![2024-11-11_17 20 25](https://github.com/user-attachments/assets/c0c945ba-6d4f-42f4-b87f-71eb09bb68fd)



tl;dr

https://github.com/user-attachments/assets/4685653d-0541-40f6-afea-477b2b22460d

# Proposal to Change Creative Buttons Texture
The creative buttons texture has haunted me for years in fabric-api. Why?  
They're blue. The entire interface for buttons hasn't been blue in over 6 years.  
Additionally they have too much contrast, sticking out like a sore thumb in the user interface.

# How do we fix it?
We have a great example of something interactable in the creative menu already, the scroll bar.
It has an inactive state and an active state:  
active:  
![scroll_enabled](https://github.com/user-attachments/assets/d6f31c7a-6320-4033-b24a-ab375a7bf7fd)
inactive:  
![scroll_disabled](https://github.com/user-attachments/assets/98d8757a-69c5-47f5-b6a2-c49f08413955)

# Extrapolating the Features
Notice that the creative scroll bar is in an indented section where it can move freely.  
This creates a contrast strong enough to differentiate it from the rest of the menu.  
The "grip" lines horizontal on the bar also changes color depending on if you can or cannot scroll.    

# Using the Features
Including the indent to seperate it from the rest of the UI helps the contrast to make it visible, without making it stand out visually from the slots and the scroll bar.
Using the pallette of the scroll bar in its active and inactive state for the pallette in the buttons being active and inactive makes it more flush with the design and not stick out.

# Result:
Default appearance:
![default_appearance](https://github.com/user-attachments/assets/ad87dff2-9ce6-464d-ba68-c6d534ae3c1a)
Hovering on right:
![hover_right](https://github.com/user-attachments/assets/d6738aad-176c-40ea-9038-d6c7fc14547d)
Hovering on left:
![hover_left](https://github.com/user-attachments/assets/67f25f40-f99c-4fcd-bef9-a832e9d44fae)
Disabled left:
![disabled_left](https://github.com/user-attachments/assets/ab416aa9-51ec-447e-b2c3-26ffe14bb9d6)
Disabled right:
![disabled_right](https://github.com/user-attachments/assets/b22637f5-76d0-4966-b569-cf69fd7cf66a)

# Full screenshot view of change
![2024-11-10_02 40 15](https://github.com/user-attachments/assets/daa48b8e-e0d0-48f0-b994-bdd29d5cfff2)

# Previous appearance for reference:
![2024-11-10_02 50 22](https://github.com/user-attachments/assets/ec6b040c-ea81-486c-8560-c45011a8419d)

# Small Feats
Reduced color pallete allows for greater compression  
5,512 bytes -> 382 bytes  
14x decrease in size  

Jar size improvement:  
1,992,580 -> 1,989,946  
0.13% total jar size improvement  
(small things matter LOL)  
detail: compiled on arch linux with Temurin-21.0.5+11-LTS   

# Footnote additional details  
The arrow is made white instead of the button outline because it becomes too noisy in a small area.  
Minecraft 1.15 removed blue backgrounds from buttons in Minecraft, and it came out in 2019, 6 years ago. 
Previous attempts, which the textures were worse rip- #2740, #2746

Laxystem commented:
I believe it makes a lot more sense to have secondary buttons there instead of primary ones.

In-game UI mostly doesn't have primary buttons (except for the `p` menu - but that one uses them as tabs, which means it's outdated anyway).

matthewperiut commented:
Another reference, back in 1.14, where hovered buttons were still blue this texture was awesome!
<img width="399" alt="Screenshot 2024-11-11 at 1 45 55 PM" src="https://github.com/user-attachments/assets/1af609a7-761a-4abf-a50c-ec0c969c79f5">
but nowadays it really doesn't make much sense with the colors chosen to be inside the existing gui as-is

embeddedt commented:
I like the new look, but the right edge of the buttons doesn't align with the right edge of the scrollbar. To be fair this was also an issue with the previous implementation, but it seems more obvious now.

matthewperiut commented:
Does this look better?
![image](https://github.com/user-attachments/assets/499b245a-9200-40d8-9457-43d4786dc8a5)
Alternatively, this?
![image](https://github.com/user-attachments/assets/8db3e84b-89c6-42f8-ab5a-00c10bda7538)

embeddedt commented:
I like the first revision, yes.

matthewperiut commented:
![image](https://github.com/user-attachments/assets/05abf95f-29b4-40ab-befe-9a3546e3b62a)
I further modified it to have the disabled button be a bit nicer

IMB11 commented:
I think this looks great :+1:

modmuss50 reviewed (approved):
Im happy with this, its a definite improvement. If anyone thinks they can be done better again please let it be known as I dont plan on touching this yet agian.

ghost commented:
Only issue I can think of is that by trying to fit into the vanilla scheme here the contrast on the active buttons end up being low.

That may not be an issue due to the physical location but I believe it is something that should still be thought about.

matthewperiut commented:
I just found a last minute thing, surprised it wasn't seen earlier by past me, but don't merge yet!
<img width="477" alt="image" src="https://github.com/user-attachments/assets/d4a06fa0-996d-4a69-a07a-670d7592f177" />

matthewperiut commented:
Alright, I'm thinking about the solutions, and here are the options,
1. Move the buttons back to their usual location
2. Shorten the search bar
3. Move the search bar

I'm not really sure what to do, but I think probably move the buttons back

embeddedt commented:
I actually think shortening the search bar would look better.

Laxystem commented:
@matthewperiut there's a fourth option, replace the outline with a normal cell border on the search tab.

IMB11 commented:
Shortening the search bar would be perfectly acceptable in my opinion, users won't notice and it's aesthetically pleasing.

Syst3ms reviewed (changes requested):
Shorten the search bar, as per consensus. Making a review of this so it doesn't get forgotten when last call ends.

matthewperiut commented:
Yeah sorry guys work got busy, let me try and get this finished

matthewperiut commented:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/ddd451ce-1144-4490-8ef9-acf8193658f8" />
getting there guys...

matthewperiut commented:
Ok, now it is ready (I think)

matthewperiut commented on fabric-item-group-api-v1/src/client/java/net/fabricmc/fabric/mixin/itemgroup/client/CreativeInventoryScreenMixin.java:
```diff
@@ -68,11 +73,11 @@ private void updateSelection() {
 	}
 
-	@Inject(method = "init", at = @At("RETURN"))
+	@Inject(method = "init", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/widget/TextFieldWidget;setEditableColor(I)V", shift = At.Shift.AFTER))
```
This was changed to stop creating creative buttons when in survival mode, incase anyone asks the purpose

matthewperiut commented:
OK I'm taking my grubby hands off this code and various like OO this could be betters and handing it off to just be reviewed, do keep in mind that I tried to keep my code as minimally invasive as possible for the changes that I made. Feel free to ask questions for my decisions. This is what the search tab looks like now
<img width="401" alt="image" src="https://github.com/user-attachments/assets/48d9b1b7-cfae-4345-8c0c-422ed3cc00b4" />

IMB11 commented:
Looks great IMO :+1:

matthewperiut commented:
Me thinking I'm done, having random thoughts of what ifs and how to fix them later... guys maybe its truly ready now.

IMB11 commented on fabric-item-group-api-v1/src/main/resources/assets/fabric/textures/gui/creative_buttons.png:
Does this PNG file need to be this big? (256x256 is a bit overkill)

matthewperiut commented:
I decided not to mess with the png size as that is how it was originally, I am truly solely trying to get the design of the button to be different and making sure the code I add isn't inefficient. But yes, in the current version widgets are usually just small images with a different image loading system.

cassiancc reviewed (changes requested):
Before this is merged, I'd like the old textures to be added to Programmer Art, so that Fabric can continue to match the older blue button style when that resource pack is enabled. I've gone ahead and set this up as a PR to matthewperiut's fork as https://github.com/matthewperiut/fabric/pull/1.

matthewperiut commented:
I feel that the user base for the addition of keeping the old textures in programmer art is a little too niche where it should just be an additional external resourcepack outside of fabric api. I will go ahead and merge it if @Syst3ms or @modmuss50 want me to though.

cassiancc commented:
There seems to be approval in the Discord. Mod Menu also ships blue buttons for [Programmer Art ](https://github.com/TerraformersMC/ModMenu/blob/1.21.2/src/main/resources/programmer_art/assets/modmenu/textures/gui/configure_button.png)and High Contrast (which I'll come back to at a later point), as they are both built-in resource packs.
![image](https://github.com/user-attachments/assets/c56d72d8-b5a4-4907-b66a-0bb824d629a7)

modmuss50 commented:
Yes, I think adding back the old textures to programmer art is a great idea. Lets do that.

I know this PR has been in last call for a very long time, but I think its been worth it to make sure that everyone can provide their input. I want to make sure we are all 100% (or as close as) happy with this before we merge it. Id rarther not come back to this anytime soon.

matthewperiut commented:
The people have spoken, its merged

matthewperiut commented:
Adding the programmer art textures has an unintended consequence of two pixels remaining that weren't masked over because in the new texture they are covered. See picture
<img width="102" alt="Screenshot 2025-01-06 at 12 21 01 PM" src="https://github.com/user-attachments/assets/7b01930e-4b24-4c6c-b90d-8174b43d79a6" />
The top left and bottom left pixels should not be there as they are a remnant of the previous search bar location pixels and were hidden before, so this next commit is to accommodate for the programmer art without this artifact, with this result:
<img width="125" alt="Screenshot 2025-01-06 at 12 20 11 PM" src="https://github.com/user-attachments/assets/cfeabf8e-9008-4c18-bc85-3ab28f174b76" />

modmuss50 commented on fabric-item-group-api-v1/src/client/java/net/fabricmc/fabric/mixin/itemgroup/client/CreativeInventoryScreenMixin.java:
```diff
@@ -197,4 +204,38 @@ public boolean setSelectedItemGroup(ItemGroup itemGroup) {
+			)
+	)
+	private TextFieldWidget shortenSearchBarTextField(TextRenderer textRenderer, int x, int y, int width, int height, Text text, Operation<TextFieldWidget> original) {
+		// Modify only the available text space width to be 5 pixels shorter
+		if (hasAdditionalPages()) {
```
No, I am strongly against changing any vanilla UI element like this, we shouldnt make something smaller, why does this need to be done?

matthewperiut replied:
A while ago someone said that the location of the buttons should be further to the left, and I moved it and people said its better moved in line with the search bar. Then I noticed the search bar goes in where the buttons go, so I asked if I should move back the buttons or shorten the search bar or move the search bar, and @Syst3ms requested to shorten the search bar basically based off consensus, which seemingly most people agreed.

This specific part of the code doesn't let the text in the search bar go too far to the right.

Another part renders the end of the search bar a little to the left.

What do you suggest I do?

matthewperiut replied:
Without this search bar change with the button locations as-is it appears like this:
![image](https://github.com/user-attachments/assets/2f004485-69d3-459c-b6b5-f8d95e19ef5d)

matthewperiut replied:
Additionally the search bar will keep going just like normal, because it allows for text longer than the length of the given area, with mojang's code making it look nice in the process

cassiancc replied:
Fabric is already changing the vanillla UI by adding buttons, is making sure our buttons fit a concern? These changes will only be made if additional Creative groups have been added by a mod.

matthewperiut replied:
I removed the wrap operation from the code to avoid modifying the element.

matthewperiut commented:
![2025-01-07_12 01 39](https://github.com/user-attachments/assets/afa3f035-4df5-4309-b2d0-149d99c5b805)
Would this be preferable?

Laxystem commented:
@matthewperiut that's my suggestion! and I, for one, like it a lot better.

matthewperiut commented:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/e52173bb-528f-457e-b74a-0007355cc6eb" /> 

I removed the wrap element, and I adjusted the positioning of everything to still make it look nice, but not exactly perfectly aligned, but this removes an invasive modification.

matthewperiut commented:
I changed the size of the buttons, the search bar is not drawn any different in this iteration, as the code has been removed, only the buttons are modified.
<img width="418" alt="image" src="https://github.com/user-attachments/assets/90db1b6c-da59-4673-b1cf-88e9913e570e" />
<img width="397" alt="image" src="https://github.com/user-attachments/assets/6b2ba0c3-88bb-4616-b80d-ab73f103e3ea" />

matthewperiut commented:
What it looks like now, as of re-requesting review
Programmer art needed a slight modification
<img width="404" alt="image" src="https://github.com/user-attachments/assets/58b82c32-b76d-4a2c-9be4-a39329a3198e" />  
<img width="401" alt="image" src="https://github.com/user-attachments/assets/b71d2d95-895e-48b0-b256-a9ae2643afb3" />  
Here's various appearances of the regular button
<img width="400" alt="image" src="https://github.com/user-attachments/assets/7098faaf-90ce-4fdc-9292-2a9a1e7b9e43" />
<img width="401" alt="image" src="https://github.com/user-attachments/assets/0aac336e-499a-424c-9cd3-09b5a41c0240" />
<img width="403" alt="image" src="https://github.com/user-attachments/assets/3c6a41f1-b178-4af9-887a-d7c39b9491ce" />
<img width="407" alt="image" src="https://github.com/user-attachments/assets/3cdc195c-8e1c-4296-a3b9-69d00263d08f" />

williamistGitHub commented:
Hey, I really like the look of these new buttons (imo they're definitely better than the old ones) but I wanted to make a suggestion about the look of them when you hover them. The old buttons change the color of the background when you hover them, but the new buttons only change the color of the arrow which makes it less obvious that something's changed (especially if a mouse pointer is covering up some of it, it's a small button). My idea to improve this is to add a white overlay to the hovered state of the buttons, so that pretty much every pixel changes color noticeably on hover.

Here's a mockup I did of the idea, I think it works well:

![image](https://github.com/user-attachments/assets/ed6a7fff-ebcd-445b-b715-6d16d3fe3f9e)

The white overlay in that example is at 40% opacity, which I think isn't too low to be not noticeable, but not too high as to make the button unrecognizable.

Thanks again for making these new buttons by the way, they look so good.

Laxystem commented:
@williamistGitHub while I agree with the concept, I think the whitening should be done in the traditional minecraft way, outlines only. Having a white overlay on the entire thing looks like my screen's damaged.

matthewperiut commented:
The indentation is being lightened, which causes it to not match the other indents on the page. Additionally whitening around the edge is something I tried but it doesn't suit this format. I've decided that the way I have it is the best way to do it unless someone has a better idea and can send a screenshot, I think it's good as-is.

IMB11 commented:
Is this going to get merged anytime soon? I think this is good as-is as well.

modmuss50 committed:
Change Creative Buttons Texture

Co-authored-by: Matthew Periut <matthewperiut@gmail.com>
